### ISSUE TITLE
Training-complete panel always shows "Loss: 0.0000 -> 0.0000" when total steps < logging_steps

### ISSUE THREAD

**Issue by Feathersabound** [MakazhanAlpamys/Soup#899]
## Summary

The "Training Complete!" rich panel printed at the end of `soup train` always shows `Loss: 0.0000 -> 0.0000` when the run has fewer total steps than `training.logging_steps` (default `10`) — even though the run had a real, non-zero loss.

## Repro

```bash
soup quickstart --dry-run -y -o /tmp/soup-demo
soup train --config /tmp/soup-demo/quickstart_soup.yaml --yes
```

This trains TinyLlama-1.1B-Chat-v1.0 with LoRA SFT on 18 samples for 1 epoch → 5 total steps (`5 < logging_steps=10`).

**Actual trainer output (correct):**
```
{'train_runtime': '4183', 'train_samples_per_second': '0.004', 'train_steps_per_second': '0.001', 'train_loss': '2.383', 'entropy': '1.117', 'num_tokens': '693', 'mean_token_accuracy': '0.6502', 'epoch': '1'}
```

**"Training Complete!" panel (wrong):**
```
╭───────────────────────────── Training Complete! ─────────────────────────────╮
│ Loss: 0.0000 -> 0.0000                                                       │
│ Duration: 1h 9m                                                              │
...
```

## Root cause

- The panel is rendered at `src/soup_cli/commands/train.py:1599`:
  ```python
  f"Loss: [bold]{result['initial_loss']:.4f} -> {result['final_loss']:.4f}[/]\n"
  ```
- `initial_loss`/`final_loss` are computed identically across essentially every trainer backend (confirmed in `sft.py:2003-2012`; the same pattern also appears in `dpo.py`, `orpo.py`, `grpo.py`, `kto.py`, `bco.py`, `ipo.py`, `simpo.py`, `online_dpo.py`, `reward_model.py`, `classifier.py`, `embedding.py`, `asr.py`, `mole_routing.py`, `pretrain.py`, `distill.py`):
  ```python
  logs = self.trainer.state.log_history
  train_losses = [entry["loss"] for entry in logs if "loss" in entry]
  ...
  "initial_loss": train_losses[0] if train_losses else 0,
  "final_loss": train_losses[-1] if train_losses else 0,
  ```
- HF `Trainer` only appends an intermediate `log_history` entry keyed `"loss"` every `logging_steps` (default `10`, see `config/schema.py:1235`). When the run has fewer total steps than that, no such entry is ever created — the only `log_history` entry is the final summary, which uses the key `"train_loss"` instead of `"loss"`.
- Because the filter is `if "loss" in entry` (a literal key match, not a substring/prefix check it happens to also match), it never matches that summary entry, so `train_losses` ends up empty and both fields silently fall back to `0`.

## Impact

Cosmetic only — the actual checkpoint/adapter and the logged `train_loss` metric are correct; only the CLI's own completion-panel summary is wrong. But it's a bad first impression for exactly the workflow (`soup quickstart`) meant to be the friendliest onboarding path, and it's a systemic copy-pasted pattern across every trainer backend, not a one-off.

## Suggested fix direction

In each trainer's summary-building code, fall back to `entry.get("train_loss")` (or check for either key) when no `"loss"`-keyed entries exist, so short runs (`total_steps < logging_steps`) still report a real value.

## Environment

- Soup v0.74.0 (fresh `pip install -e ".[dev]"`)
- Python 3.12.14, CPU-only training
- torch 2.14.0, transformers 5.17.0, trl 0.29.1, peft 0.20.0

**Comment by MakazhanAlpamys** [comment]
Confirmed, and thank you — you did the work I would otherwise have had to do: the root cause is right, the file:line references are right, and enumerating the sixteen trainers that share the pattern is what makes this worth fixing once rather than sixteen times.

The mechanism is exactly as you describe. `if "loss" in entry` on a `dict` tests **keys**, HF `Trainer` writes intermediate entries keyed `loss` only every `logging_steps`, and the final summary entry uses `train_loss` instead. Fewer steps than `logging_steps` therefore means `train_losses == []`, and both ends of the pair silently become `0`. A `0.0000 -> 0.0000` that is indistinguishable from a genuinely dead run is the part that makes this a bug rather than cosmetics — that is the display a user would take to an issue tracker.

**One design trap for whoever fixes it, so the fix does not become a second wrong number.** The summary entry's `train_loss` is a **mean over the whole run**, not the first observation. Falling back to it for *both* ends would print `Loss: 2.3830 -> 2.3830`, which reads as "the loss did not move" — a different false statement, and a harder one to notice. Two options that do not lie:

- print a single value with no arrow when only the run-level mean is available (`Loss: 2.3830 (mean over 5 steps; fewer steps than logging_steps=10, so no per-step history)`), or
- keep the arrow only when `len(train_losses) >= 2`, and otherwise fall back to the single-value form.

Either is fine by me; the property to hold is that the panel never shows a *delta* it does not have the data for. Whatever you pick, the test should pin both branches — a short run and a normal one — and pin that the short-run branch does not emit an arrow.

The sixteen call sites are the other half. They are verbatim copies, so please extract the extraction (a small shared helper that takes `trainer.state.log_history` and returns the pair or the single value) rather than editing the same four lines sixteen times; a seventeenth trainer added later should get the fixed behaviour for free. A repo-wide test asserting no trainer keeps a private copy would be welcome but is not a requirement.

Labelled `bug` + `good first issue` + `help wanted`. **Would you like to take it?** You have already done the diagnosis, so the implementation is the small part — say the word and I will mark it `claimed` for you. If you would rather not, it stays open for anyone.

To be precise about what I checked myself: I verified the **mechanism** by reading the code, not by running the training — `logging_steps` defaults to `10` (`config/schema.py:1235`) and `sft.py:2003-2012` is the key-filter you quote, verbatim. The step count also lines up on paper: `DEMO_DATA` is 20 rows, `val_split` takes two, `batch_size: auto` lands at 4, so 5 steps against a threshold of 10. I have not executed the hour-long quickstart run, so the panel output itself is your observation rather than mine — which it does not need to be, since the mechanism is sufficient.

**Comment by taljeon** [comment]
Yes, I'd like to take this. I'll extract the shared loss-summary logic, migrate the duplicated trainer call sites, and add tests for both the short-run single-value/no-arrow case and the normal per-step delta case. I'll keep the verification CPU/static and won't rerun the hour-long training reproduction.

**Comment by MakazhanAlpamys** [comment]
Removing the `good first issue` label — the issue is real and well-reported, the label was wrong.

The body itself says the pattern is duplicated across sixteen trainers, and PR #908 (now open against this) confirms the scale: **20 source files plus a new `trainer/loss_summary.py`**. That is a cross-cutting refactor touching every trainer wrapper in the repo, with a merge-order hazard against #919 and #923 which rewrite overlapping sets of the same files. A newcomer taking this would walk into a three-way rebase on their first contribution.

Keeping `bug` and `help wanted`. Nothing about the report changes — this is purely a labelling correction so the next person reading the tracker gets an accurate signal about size.

**Comment by MakazhanAlpamys** [comment]
Fixed on `main` by @taljeon in #908, merged as `6ff3ca2d`.

`soup train` reported a loss delta for runs too short to have one, computing it against a baseline that did not exist yet — a fabricated number presented in the same format as a real measurement, which is the failure mode that teaches people to trust the display.

The change arrived through a nine-file conflict resolution, checked hunk by hunk against both sides rather than by confirming the merge was clean; git reporting no conflict is not evidence about a resolution of that size. Verified on the merged tree with #931, which touches `trainer/sft.py` and `trainer/pretrain.py` in the same region.

Credit to @taljeon.
